src/ripple/consensus/Consensus.h

```diff
@@ -1108,6 +1135,113 @@ Consensus<Adaptor>::phaseOpen()
     }
 }
 
+template <class Adaptor>
+bool
+Consensus<Adaptor>::shouldPause() const
+{
+    auto const& parms = adaptor_.parms();
+    std::uint32_t const desynced (previousLedger_.seq() -
+        std::min(adaptor_.getValidLedgerIndex(), previousLedger_.seq()));
+    auto quorumKeys = adaptor_.getQuorumKeys();
+    auto const& quorum = quorumKeys.first;
+    auto& trustedKeys = quorumKeys.second;
+    std::size_t const totalValidators = trustedKeys.size();
+    std::size_t laggards = adaptor_.laggards(previousLedger_.seq(),
+        trustedKeys);
+
+    std::stringstream vars;
+    vars << " (working seq: " << previousLedger_.seq() << ", "
+         << "validated seq: " << adaptor_.getValidLedgerIndex() << ", "
+         << "am validator: " << adaptor_.validator() << ", "
+         << "have validated: " << adaptor_.haveValidated() << ", "
+         << "roundTime: " << result_->roundTime.read().count() << ", "
+         << "max consensus time: " << parms.ledgerMAX_CONSENSUS.count() << ", "
+         << "validators: " << totalValidators << ", "
+         << "laggards: " << laggards << ", "
+         << "offline: " << trustedKeys.size() << ", "
+         << "quorum: " << quorum << ")";
+
+    if (!desynced ||
+        !laggards ||
+        !totalValidators ||
+        !adaptor_.validator() ||
+        !adaptor_.haveValidated() ||
+        result_->roundTime.read() > parms.ledgerMAX_CONSENSUS)
+    {
+        j_.debug() << "not pausing" << vars.str();
+        return false;
+    }
+
+    bool willPause = false;
+
+    /** Maximum phase with distinct thresholds to determine how
+     *  many validators must be on our same ledger sequence number.
+     *  The threshold for the 1st (0) phase is >= the minimum number that
+     *  can achieve quorum. Threshold for the maximum phase is 100%
+     *  of all trusted validators. Progression from min to max phase is
+     *  simply linear. If there are 5 phases (maxPausePhase = 4)
+     *  and minimum quorum is 80%, then thresholds progress as follows:
+     *  0: >=80%
+     *  1: >=85%
+     *  2: >=90%
+     *  3: >=95%
+     *  4: =100%
+     */
+    constexpr static std::size_t maxPausePhase = 4;
+
+    /**
+     * No particular threshold guarantees consensus, but lower thresholds
+     * are easier to achieve than higher (but higher thresholds tend to
+     * be more likely to reach consensus). Cycle through the phases if
+     * lack of synchronization continues.
+     */
+    std::size_t const phase = desynced % (maxPausePhase + 1);
+
+    // validators that remain after the laggards() function are considered
+    // offline, and should be considered as laggards for purposes of
+    // evaluating whether the threshold for non-laggards has been reached.
+    switch (phase)
+    {
+        case 0:
+            // Laggards and offline shouldn't preclude consensus.
+            if (laggards + trustedKeys.size() > totalValidators - quorum)
```

Comment:
Also, should you just be using the quantity `laggards + trustedKeys.size()` here and below? I don't see them used independently.  When does that distinction matter?

Reply:
// Ensure that sufficient validators are known to be not lagging.
trustedKeys.size() are offline
If too many validators are offline, then we are nearly certain to not reach network consensus if we close. So pausing won't hurt, and will most likely only help.

Reply:
> Also, should you just be using the quantity `laggards + trustedKeys.size()` here and below? I don't see them used independently. When does that distinction matter?

I'm not sure I understand your question. I am using those quantities. Do you mean should assign a variable to them instead of calculating? Partly because the branching of the switch means that the value is only used a single time. Also because I think it adds clarity to the formulas to be explicit about those values and their meaning--rather than making up something else which simply means "laggards+offline".